### PR TITLE
module: prepare sink & source in bind & unbind function

### PR DESCRIPTION
**Pull request by RanderWang**
Module will update source & sink information when bind & unbind event happen.

### CODE REVIEW

**Review by marcinszkudlinski — CHANGES_REQUESTED** [review]
Don't duplicate identical functionality - pointers to sink and sources are set twice, using identical code

remove ``module_adapter_sink_src_prepare`` from module adapter - consider applying https://github.com/thesofproject/sof/commit/d0bf41b6b3df45c40dfcd30a7329e284250c3346 together with this patch

**Comment by RanderWang** [comment]
> Don't duplicate identical functionality - pointers to sink and sources are set twice, using identical code
> 
> remove `module_adapter_sink_src_prepare` from module adapter - consider applying [d0bf41b](https://github.com/thesofproject/sof/commit/d0bf41b6b3df45c40dfcd30a7329e284250c3346) together with this patch

@marcinszkudlinski  Thanks, integrate your suggestion.

**Review by marcinszkudlinski — APPROVED** [review]

**Review by kv2019i — APPROVED** [review]

File: src/audio/module_adapter/module_adapter.c
Comment (btian1) [inline src/audio/module_adapter/module_adapter.c:168]: in line 180 and 222, mod->num_of_sources and mod->num_of_sinks also re-calculated, is it possible to keep one?

Comment (RanderWang) [reply src/audio/module_adapter/module_adapter.c:168]: It is for dp_queue. I am not sure about this. @marcinszkudlinski  is working on dp queue.

File: src/include/sof/audio/module_adapter/module/modules.h
Comment (lyakh) [inline src/include/sof/audio/module_adapter/module/modules.h:70]: just for clarity: this change only affects loadable modules, for built-in modules `.bind` and `.unbind` are already set in generic.h

**Review by lyakh — APPROVED** [review]

File: src/audio/module_adapter/module_adapter.c
Comment (marcinszkudlinski) [reply src/audio/module_adapter/module_adapter.c:168]: Yes, it should be changed - creating of "shadow dpQueues" should also be moved to bind/unbind. 
It is not hurting DP processing at the moment and I don't want to delay this PR.  - so I approved the PR and will change it in separate PR

Target solution is Pipeline 2.0 where sink/src will be a main data exchanging API

**Comment by kv2019i** [comment]
This looks good to go otherwise, but the testbench failure needs to be fixed. If it's a problem in testbench, we can have a fix in separate PR, but we need a rootcause before we can merge this.

**Comment by singalsu** [comment]
The fail with testbench (and likely all IPC3) is because num_of_sources and num_of_sinks is zero when calling src_prepare().

Screenshot from 2023-12-11 16-25-18

**Comment by singalsu** [comment]
@RanderWang Does the SRC work in a normal IPC4 device with this PR? I'm not sure we have in CI test that runs a SRC pipeline. Looks that mod->num_of_sources and mod->num_of_sinks() are set after calling module_prepare() in module_adapter_dp_queue_prepare().

For testbench in module_adapter_prepare() the first if true is not taken because mod->dev->ipc_config.proc_domain is not COMP_PROCESSING_DOMAIN_DP. Then in the second else if the num_of_ arguments are zeros.

```
	if (IS_PROCESSING_MODE_SINK_SOURCE(mod) &&
	    mod->dev->ipc_config.proc_domain == COMP_PROCESSING_DOMAIN_DP)
		ret = module_adapter_dp_queue_prepare(dev);

	else if (IS_PROCESSING_MODE_SINK_SOURCE(mod) &&
		 mod->dev->ipc_config.proc_domain == COMP_PROCESSING_DOMAIN_LL)
		ret = module_prepare(mod, mod->sources, mod->num_of_sources,
				     mod->sinks, mod->num_of_sinks);

```

File: src/include/sof/audio/module_adapter/module/modules.h
Comment (RanderWang) [reply src/include/sof/audio/module_adapter/module/modules.h:70]: sure

**Comment by RanderWang** [comment]
> @RanderWang Does the SRC work in a normal IPC4 device with this PR? I'm not sure we have in CI test that runs a SRC pipeline. Looks that mod->num_of_sources and mod->num_of_sinks() are set after calling module_prepare() in module_adapter_dp_queue_prepare().
> 
> For testbench in module_adapter_prepare() the first if true is not taken because mod->dev->ipc_config.proc_domain is not COMP_PROCESSING_DOMAIN_DP. Then in the second else if the num_of_ arguments are zeros.
> 
> ```
> 	if (IS_PROCESSING_MODE_SINK_SOURCE(mod) &&
> 	    mod->dev->ipc_config.proc_domain == COMP_PROCESSING_DOMAIN_DP)
> 		ret = module_adapter_dp_queue_prepare(dev);
> 
> 	else if (IS_PROCESSING_MODE_SINK_SOURCE(mod) &&
> 		 mod->dev->ipc_config.proc_domain == COMP_PROCESSING_DOMAIN_LL)
> 		ret = module_prepare(mod, mod->sources, mod->num_of_sources,
> 				     mod->sinks, mod->num_of_sinks);
> ```

I tested SRC. but it is working LL mode not in DP mode

**Comment by RanderWang** [comment]
I also tested SRC in DP mode, but no issue found

**Review by btian1 — APPROVED** [review]

**Comment by RanderWang** [comment]
Even I simply this PR to one which just updates source & sink info in bind &unbind function but mixin & mixout will be failed.   Actually sink & source are not used by mixin & mixout.  .......

**Review by abonislawski — APPROVED** [review]

**Comment by singalsu** [comment]
> I also tested SRC in DP mode, but no issue found

Thanks, then it looks like this is only issue for IPC3 or testbench. Can you check if adding for IPC3 build some of the removed num_of calculations fixes the test issue? In the above code snippet, for the else if call before calling module_prepare().

**Comment by RanderWang** [comment]
> > I also tested SRC in DP mode, but no issue found
> 
> Thanks, then it looks like this is only issue for IPC3 or testbench. Can you check if adding for IPC3 build some of the removed num_of calculations fixes the test issue? In the above code snippet, for the else if call before calling module_prepare().

thanks, I got the same idea. IPC3 doesnt has bind & unbind

**Comment by kv2019i** [comment]
Looks better now, let's wait for the mandatory CI runs to complete and if all good, this is ready to be merged.

**Comment by kv2019i** [comment]
CI almost clean. I do see one fail here:
https://sof-ci.01.org/sofpr/PR8594/build1044/devicetest/index.html?model=TGLU_RVP_SDW-ipc4&testcase=check-signal-stop-start-playback-10

Looks like a case of https://github.com/thesofproject/sof/issues/8056 but triggered in a different test case.

**Comment by kv2019i** [comment]
@RanderWang Can you check the above fail? I browsed through history of PR tests and I don't see this failing for any other PR.

**Comment by RanderWang** [comment]
> @RanderWang Can you check the above fail? I browsed through history of PR tests and I don't see this failing for any other PR.

@kv2019i  it was a old bug hard to be reproduced. Let run it again.  
````c
[   53.980498] <err> dai_comp: dai_common_new: comp:1 0x10004 dai_new(): dai_get() failed to create DAI.
[   53.980518] <err> copier: copier_dai_create: comp:1 0x10004 failed to create dai
````

**Comment by RanderWang** [comment]
SOFCI TEST

**Comment by RanderWang** [comment]
@kv2019i  pass. It is unlucky that it can't be reproduced again.

**Pull request merged**
